File: frontend/playground/battle/calculator/vehicle/ground-reaction-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import type {VehicleCalculations, WheelCalculations} from "~/playground/data/calculations";
import type {BattleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";
import {VectorUtils} from "~/playground/utils/vector-utils";
import type {Position} from "~/playground/data/common";
import {BodyUtils} from "~/playground/utils/body-utils";

export class GroundReactionForceCalculator implements ForceCalculator<VehicleCalculations> {
  private static readonly FORCE_DESCRIPTION = 'Ground reaction'

  calculate(calculations: VehicleCalculations, battleModel: BattleModel): BodyForce[] {
    const groundReactionCoefficient = battleModel.room.specs.groundReactionCoefficient
    const groundMaxDepth = battleModel.room.specs.groundMaxDepth
    const comPosition = calculations.model.state.position
    const forces: BodyForce[] = []

    this.addWheelReaction(forces, calculations.rightWheel, comPosition, groundReactionCoefficient, groundMaxDepth)
    this.addWheelReaction(forces, calculations.leftWheel, comPosition, groundReactionCoefficient, groundMaxDepth)
    this.addHullReaction(forces, calculations, groundReactionCoefficient, groundMaxDepth)

    return forces
  }

  private addWheelReaction(
      forces: BodyForce[],
      calculations: WheelCalculations,
      comPosition: Position,
      groundReactionCoefficient: number,
      groundMaxDepth: number
  ): void {
    const groundContact = calculations.groundContact
    if (!groundContact) {
      return
    }

    const velocityNormalProjectionMagnitude = VectorUtils.getMagnitude(
        VectorUtils.projectionOfOnto(calculations.velocity!, groundContact.normal)
    )

    if (velocityNormalProjectionMagnitude > 0) {
      const depth = Math.min(groundContact.depth, groundMaxDepth)
      const force = VectorUtils.multiply(groundContact.normal,
          -velocityNormalProjectionMagnitude * depth * groundReactionCoefficient)

      forces.push(BodyForce.of(
          force,
          groundContact.position,
          comPosition,
          GroundReactionForceCalculator.FORCE_DESCRIPTION
      ))
    }
  }

  private addHullReaction(
      forces: BodyForce[],
      calculations: VehicleCalculations,
      groundReactionCoefficient: number,
      groundMaxDepth: number
  ): void {
    const groundContacts = calculations.groundContacts
    if (!groundContacts) {
      return
    }
    const comPosition = calculations.model.state.position

    groundContacts.forEach(contact => {
      const velocityNormalProjectionMagnitude = VectorUtils.getMagnitude(VectorUtils.projectionOfOnto(
          BodyUtils.getVelocityAt(calculations.model.state, contact.position), contact.normal))
      if (velocityNormalProjectionMagnitude > 0) {
        const depth = Math.min(contact.depth, groundMaxDepth)
        const force = VectorUtils.multiply(contact.normal,
            -velocityNormalProjectionMagnitude * depth * groundReactionCoefficient)

        forces.push(BodyForce.of(
            force,
            contact.position,
            comPosition,
            GroundReactionForceCalculator.FORCE_DESCRIPTION
        ))
      }
    })
  }
}

File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/box/GroundReactionForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.box;

import com.github.aadvorak.artilleryonline.battle.calculations.BodyForce;
import com.github.aadvorak.artilleryonline.battle.calculations.BoxCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.ForceCalculator;
import com.github.aadvorak.artilleryonline.battle.calculator.utils.ForceCalculatorUtils;
import com.github.aadvorak.artilleryonline.battle.config.BoxConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.BoxModel;
import com.github.aadvorak.artilleryonline.battle.precalc.BoxPreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.BoxSpecs;
import com.github.aadvorak.artilleryonline.battle.state.BoxState;

import java.util.ArrayList;
import java.util.List;

public class GroundReactionForceCalculator implements ForceCalculator<
        BoxSpecs,
        BoxPreCalc,
        BoxConfig,
        BoxState,
        BoxModel,
        BoxCalculations> {

    private static final String FORCE_DESCRIPTION = "Ground reaction";

    @Override
    public List<BodyForce> calculate(BoxCalculations calculations, BattleModel battleModel) {
        var groundReactionCoefficient = battleModel.getRoom().getSpecs().getGroundReactionCoefficient();
        var groundMaxDepth = battleModel.getRoom().getSpecs().getGroundMaxDepth();
        var forces = new ArrayList<BodyForce>();
        if (calculations.getGroundContacts() != null) {
            calculations.getGroundContacts().forEach(contact -> ForceCalculatorUtils.addReaction(forces,
                    calculations, contact, groundMaxDepth, groundReactionCoefficient, FORCE_DESCRIPTION));
        }
        return forces;
    }
}
